File: brixen-java/src/test/java/org/brixen/test/bean/HoverControlBeanTest.java
package org.brixen.test.bean;

import org.brixen.bean.HoverControlBean;
import org.brixen.bean.HoverControlBeanImpl;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.openqa.selenium.WebElement;
import org.testng.Assert;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import static org.mockito.Mockito.when;
import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertNotEquals;
import static org.testng.Assert.assertTrue;

@Test(groups = {"LoadableBeanTestGroup"})
public class HoverControlBeanTest {
    private @Mock WebElement mockContainerElementOne;
    private @Mock WebElement mockContainerElementTwo;
    private @Mock WebElement mockUnhoverElementOne;
    private @Mock WebElement mockUnhoverElementTwo;

    @BeforeMethod
    public void setUp() {
        MockitoAnnotations.initMocks(this);
    }

    @SuppressWarnings("ConstantConditions")
    @Test(groups = {"ExpectedExceptionsTestGroup", "BeanExpectedExceptionsTestGroup",
            "HoverControlBeanExpectedExceptionsTestGroup"},
            expectedExceptions = { NullPointerException.class }
    )
    public void shouldThrowExceptionForNullUnhoverElement() {
        final HoverControlBean bean = new HoverControlBeanImpl();
        bean.setUnhoverElement(null);
    }

    @Test(groups = {"EqualityNegativeTestGroup", "BeanEqualityNegativeTestGroup",
            "HoverControlBeanEqualityNegativeTestGroup"})
    public void shouldNotBeEqual() {
        HoverControlBean beanOne = new HoverControlBeanImpl();
        HoverControlBean beanTwo = new HoverControlBeanImpl();

        beanOne.setContentContainer(mockContainerElementOne);
        beanTwo.setContentContainer(mockContainerElementTwo);

        assertNotEquals(beanOne, beanTwo, "Beans with different container element references should not be equal");

        beanTwo.setContentContainer(mockContainerElementOne);
        beanOne.setUnhoverElement(mockUnhoverElementOne);

        assertNotEquals(beanOne, beanTwo, "Beans with different unhover element references should not be equal");

        beanTwo.setUnhoverElement(mockUnhoverElementTwo);

        assertNotEquals(beanOne, beanTwo, "Beans with different unhover element references should not be equal");

        beanTwo.setUnhoverElement(mockUnhoverElementOne);
        beanOne.setHoverWithJavascript(true);

        assertNotEquals(beanOne, beanTwo, "Beans with different JavaScript hover workaround values should not be " +
                "equal");

        beanTwo.setHoverWithJavascript(true);
        beanOne.setUnhoverWithJavascript(true);

        assertNotEquals(beanOne, beanTwo, "Beans with different JavaScript unhover workaround values should not be " +
                "equal");

        beanTwo.setUnhoverWithJavascript(true);
        beanOne.setClickInsteadOfHover(true);

        assertNotEquals(beanOne, beanTwo, "Beans with different click action hover workaround values should not be " +
                "equal");

        beanTwo.setClickInsteadOfHover(true);
        beanOne.setClickWithJavascriptInsteadOfHover(true);

        assertNotEquals(beanOne, beanTwo, "Beans with different JavaScript click action hover workaround values " +
                "should not be equal");

        beanTwo.setClickWithJavascriptInsteadOfHover(true);
        beanOne.setUnhoverWithClickInstead(true);

        assertNotEquals(beanOne, beanTwo, "Beans with different click action unhover workaround values should not be " +
                "equal");

        beanTwo.setUnhoverWithClickInstead(true);
        beanOne.setUnhoverWithJavascriptClickInstead(true);

        assertNotEquals(beanOne, beanTwo, "Beans with different JavaScript click action unhover workaround values " +
                "should not be equal");
    }

    @Test(groups = {"EqualityPositiveTestGroup", "BeanEqualityPositiveTestGroup",
            "HoverControlBeanEqualityPositiveTestGroup"}
    )
    public void shouldBeEqual() {
        HoverControlBean beanOne = new HoverControlBeanImpl();
        HoverControlBean beanTwo = new HoverControlBeanImpl();

        assertEquals(beanOne, beanTwo, "Two newly constructed beans should be equal before any setters are invoked");

        beanOne.setContentContainer(mockContainerElementOne);
        beanTwo.setContentContainer(mockContainerElementOne);

        assertEquals(beanOne, beanTwo, "Two beans should be equal if they have the same container element reference " +
                "and same default JavaScript click workaround value");

        beanOne.setUnhoverElement(mockUnhoverElementOne);
        beanTwo.setUnhoverElement(mockUnhoverElementOne);

        assertEquals(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " +
                "element references and same default workaround values");

        beanOne.setHoverWithJavascript(true);
        beanTwo.setHoverWithJavascript(true);

        assertEquals(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " +
                "element references and same non-default JavaScript hover workaround value");

        beanOne.setHoverWithJavascript(false);
        beanTwo.setHoverWithJavascript(false);
        beanOne.setUnhoverWithJavascript(true);
        beanTwo.setUnhoverWithJavascript(true);

        assertEquals(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " +
                "element references and same non-default JavaScript unhover workaround value");

        beanOne.setUnhoverWithJavascript(false);
        beanTwo.setUnhoverWithJavascript(false);
        beanOne.setClickInsteadOfHover(true);
        beanTwo.setClickInsteadOfHover(true);

        assertEquals(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " +
                "element references and same non-default click action hover workaround value");

        beanOne.setClickInsteadOfHover(false);
        beanTwo.setClickInsteadOfHover(false);
        beanOne.setClickWithJavascriptInsteadOfHover(true);
        beanTwo.setClickWithJavascriptInsteadOfHover(true);

        assertEquals(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " +
                "element references and same non-default JavaScript click action hover workaround value");

        beanOne.setClickWithJavascriptInsteadOfHover(false);
        beanTwo.setClickWithJavascriptInsteadOfHover(false);
        beanOne.setUnhoverWithJavascript(true);
        beanTwo.setUnhoverWithJavascript(true);

        assertEquals(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " +
                "element references and same non-default click action unhover workaround value");

        beanOne.setUnhoverWithJavascript(false);
        beanTwo.setUnhoverWithJavascript(false);
        beanOne.setUnhoverWithClickInstead(true);
        beanTwo.setUnhoverWithClickInstead(true);

        assertEquals(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " +
                "element references and same non-default click action unhover workaround value");

        beanOne.setUnhoverWithClickInstead(false);
        beanTwo.setUnhoverWithClickInstead(false);
        beanOne.setUnhoverWithJavascriptClickInstead(true);
        beanTwo.setUnhoverWithJavascriptClickInstead(true);

        assertEquals(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " +
                "element references and same non-default JavaScript click action unhover workaround value");

        beanOne = new HoverControlBeanImpl();
        beanTwo = new HoverControlBeanImpl();

        beanOne.setHoverWithJavascript(true);
        beanTwo.setHoverWithJavascript(true);
        beanOne.setUnhoverWithJavascript(true);
        beanTwo.setUnhoverWithJavascript(true);
        beanOne.setClickInsteadOfHover(true);
        beanTwo.setClickInsteadOfHover(true);
        beanOne.setClickWithJavascriptInsteadOfHover(true);
        beanTwo.setClickWithJavascriptInsteadOfHover(true);
        beanOne.setUnhoverWithClickInstead(true);
        beanTwo.setUnhoverWithClickInstead(true);
        beanOne.setUnhoverWithJavascriptClickInstead(true);
        beanTwo.setUnhoverWithJavascriptClickInstead(true);

        assertEquals(beanOne, beanTwo, "Two beans should be equal if they have null container and unhover " +
                 "element references and same non-default JavaScript hover workaround values");
    }

    @Test(groups = {"ToStringCallsSuperTestGroup", "BeanToStringCallsSuperTestGroup",
            "HoverControlBeanToStringCallsSuperTestGroup"}
    )
    public void shouldCallSuperForToString() {
        final HoverControlBean bean = new HoverControlBeanImpl();

        when(mockContainerElementOne.toString()).thenReturn("Mock Container WebElement");
        bean.setContentContainer(mockContainerElementOne);

        assertTrue(bean.toString().contains("contentContainer=Mock Container WebElement"), "toString() for " +
                "HoverControlBeanImpl does not include the content container field in ContainerBeanImpl: " +
                bean.toString());
    }

    @Test(groups = {"HashCodeCallsSuperTestGroup", "BeanHashCodeCallsSuperTestGroup",
            "HoverControlBeanHashCodeCallsSuperTestGroup"}, dependsOnMethods = {"shouldCallSuperForToString"}
    )
    public void shouldCallSuperForHashCode() {
        final HoverControlBean bean = new HoverControlBeanImpl();
        final HoverControlBean beanToCompare = new HoverControlBeanImpl();

        assertEquals(bean.hashCode(), beanToCompare.hashCode(), "Hash codes for bean which have not had setters " +
                "called should be equal, but are not: " + bean.toString() + ", " + beanToCompare.toString());

        bean.setContentContainer(mockContainerElementOne);
        assertNotEquals(bean.hashCode(), beanToCompare.hashCode(), "Hash codes for bean which have different values " +
                "for their content container fields should not be equal, but are: " + bean.toString() + ", " +
                beanToCompare.toString());

        beanToCompare.setContentContainer(mockContainerElementOne);
        assertEquals(bean.hashCode(), beanToCompare.hashCode(), "Hash codes for bean which have the same content " +
                "container should be equal, but are not: " + bean.toString() + ", " + beanToCompare.toString());
    }

    @Test(groups = {"EqualsCallsSuperTestGroup", "BeanEqualsCallsSuperTestGroup",
            "HoverControlBeanEqualsCallsSuperTestGroup"},
            dependsOnMethods = {"shouldCallSuperForToString", "shouldCallSuperForHashCode"}
    )
    public void shouldCallSuperForEquals() {
        final HoverControlBean bean = new HoverControlBeanImpl();
        final HoverControlBean beanToCompare = new HoverControlBeanImpl();

        assertEquals(bean, beanToCompare, "Equals method should return true, but did not: " + bean.toString() + ", " +
                beanToCompare.toString());

        bean.setContentContainer(mockContainerElementOne);
        assertNotEquals(bean, beanToCompare, "Equals method should return false, but did not: " + bean.toString() +
                ", " + beanToCompare.toString());

        beanToCompare.setContentContainer(mockContainerElementOne);
        assertEquals(bean, beanToCompare, "Equals method should return true, but did not: " + bean.toString() + ", " +
                beanToCompare.toString());
    }

    @Test
    public void shouldReturnTrueForIsHoverControl() {
        final HoverControlBean bean = new HoverControlBeanImpl();
        Assert.assertTrue(bean.isHoverControl());
    }

    @Test
    public void shouldReturnFalseForIsClickControl() {
        final HoverControlBean bean = new HoverControlBeanImpl();
        Assert.assertFalse(bean.isClickControl());
    }

    @Test
    public void shouldReturnFalseForIsHoverAndClickControl() {
        final HoverControlBean bean = new HoverControlBeanImpl();
        Assert.assertFalse(bean.isHoverAndClickControl());
    }
}

File: brixen-dotnet/test/bean/HoverControlBeanTest.cs
using System;
using NUnit.Framework;
using OpenQA.Selenium;
using Moq;

namespace Org.Brixen.Bean.Tests {

	[TestFixture]
	[Category("HoverControlBeanTestGroup")]
	public class HoverControlBeanTest {

		[Test]
		[Category("ExpectedExceptionsTestGroup")]
		[Category("BeanExpectedExceptionsTestGroup")]
		[Category("HoverControlBeanExpectedExceptionsTestGroup")]
		[ExpectedException("System.ArgumentNullException", ExpectedMessage = "Cannot invoke setter for " + 
			"UnhoverElement property with a null parameter", MatchType=MessageMatch.Regex)]
		public void shouldThrowExceptionForNullUnhoverElement() {
			IHoverControlBean bean = new HoverControlBean();
			bean.UnhoverElement = null;
		}

		[Test]
		[Category("EqualityNegativeTestGroup")]
		[Category("BeanEqualityNegativeTestGroup")]
		[Category("HoverControlBeanEqualityNegativeTestGroup")]
		public void ShouldNotBeEqual() {
			IHoverControlBean beanOne = new HoverControlBean();
			IHoverControlBean beanTwo = new HoverControlBean();

			Mock<IWebElement> mockContainerElementOne = new Mock<IWebElement>();
			Mock<IWebElement> mockContainerElementTwo = new Mock<IWebElement>();

			beanOne.ContentContainer = mockContainerElementOne.Object;
			beanTwo.ContentContainer = mockContainerElementTwo.Object;

			Assert.AreNotEqual(beanOne, beanTwo, "Beans with different container element references should not be " + 
				"equal");

			Mock<IWebElement> mockUnhoverElementOne = new Mock<IWebElement>();
			Mock<IWebElement> mockUnhoverElementTwo = new Mock<IWebElement>();

			beanTwo.ContentContainer = mockContainerElementOne.Object;
			beanOne.UnhoverElement= mockUnhoverElementOne.Object;

			Assert.AreNotEqual(beanOne, beanTwo, "Beans with different unhover element references should not be equal");

			beanTwo.UnhoverElement = mockUnhoverElementTwo.Object;

			Assert.AreNotEqual(beanOne, beanTwo, "Beans with different unhover element references should not be equal");

			beanTwo.UnhoverElement = mockUnhoverElementOne.Object;
			beanOne.HoverWithJavascript = true;

			Assert.AreNotEqual(beanOne, beanTwo, "Beans with different JavaScript hover workaround values should not " + 
				"be equal");

			beanTwo.HoverWithJavascript = true;
			beanOne.UnhoverWithJavascript = true;

			Assert.AreNotEqual(beanOne, beanTwo, "Beans with different JavaScript unhover workaround values should " + 
				"not be equal");

			beanTwo.UnhoverWithJavascript = true;
			beanOne.ClickInsteadOfHover = true;

			Assert.AreNotEqual(beanOne, beanTwo, "Beans with different click action hover workaround values should " + 
				"not be equal");

			beanTwo.ClickInsteadOfHover = true;
			beanOne.ClickWithJavascriptInsteadOfHover = true;

			Assert.AreNotEqual(beanOne, beanTwo, "Beans with different JavaScript click action hover workaround " + 
				"values should not be equal");

			beanTwo.ClickWithJavascriptInsteadOfHover = true;
			beanOne.UnhoverWithClickInstead = true;

			Assert.AreNotEqual(beanOne, beanTwo, "Beans with different click action unhover workaround values should " + 
				"not be equal");

			beanTwo.UnhoverWithClickInstead = true;
			beanOne.UnhoverWithJavascriptClickInstead = true;

			Assert.AreNotEqual(beanOne, beanTwo, "Beans with different JavaScript click action unhover workaround " + 
				"values should not be equal");
		}

		[Test]
		[Category("EqualityPositiveTestGroup")]
		[Category("BeanEqualityPositiveTestGroup")]
		[Category("HoverControlBeanEqualityPositiveTestGroup")]
		public void ShouldBeEqual() {
			IHoverControlBean beanOne = new HoverControlBean();
			IHoverControlBean beanTwo = new HoverControlBean();

			Assert.AreEqual(beanOne, beanTwo, "Two newly constructed beans should be equal before any setters are " + 
				"invoked");

			Mock<IWebElement> mockContainerElement = new Mock<IWebElement>();

			beanOne.ContentContainer = mockContainerElement.Object;
			beanTwo.ContentContainer = mockContainerElement.Object;

			Assert.AreEqual(beanOne, beanTwo, "Two beans should be equal if they have the same container element " + 
				"reference and same default JavaScript click workaround value");
			
			Mock<IWebElement> mockUnhoverElement = new Mock<IWebElement>();

			beanOne.UnhoverElement = mockUnhoverElement.Object;
			beanTwo.UnhoverElement = mockUnhoverElement.Object;

			Assert.AreEqual(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " +
				"element references and same default workaround values");

			beanOne.HoverWithJavascript = true;
			beanTwo.HoverWithJavascript = true;

			Assert.AreEqual(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " + 
				"element references and same non-default JavaScript hover workaround value");

			beanOne.HoverWithJavascript = false;
			beanTwo.HoverWithJavascript = false;
			beanOne.UnhoverWithJavascript = true;
			beanTwo.UnhoverWithJavascript = true;

			Assert.AreEqual(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " + 
				"element references and same non-default JavaScript unhover workaround value");

			beanOne.UnhoverWithJavascript = false;
			beanTwo.UnhoverWithJavascript = false;
			beanOne.ClickInsteadOfHover = true;
			beanTwo.ClickInsteadOfHover = true;

			Assert.AreEqual(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " + 
				"element references and same non-default click action hover workaround value");

			beanOne.ClickInsteadOfHover = false;
			beanTwo.ClickInsteadOfHover = false;
			beanOne.ClickWithJavascriptInsteadOfHover = true;
			beanTwo.ClickWithJavascriptInsteadOfHover = true;

			Assert.AreEqual(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " + 
				"element references and same non-default JavaScript click action hover workaround value");

			beanOne.ClickWithJavascriptInsteadOfHover = false;
			beanTwo.ClickWithJavascriptInsteadOfHover = false;
			beanOne.UnhoverWithJavascript = true;
			beanTwo.UnhoverWithJavascript = true;

			Assert.AreEqual(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " + 
				"element references and same non-default click action unhover workaround value");

			beanOne.UnhoverWithJavascript = false;
			beanTwo.UnhoverWithJavascript = false;
			beanOne.UnhoverWithClickInstead = true;
			beanTwo.UnhoverWithClickInstead = true;

			Assert.AreEqual(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " + 
				"element references and same non-default click action unhover workaround value");

			beanOne.UnhoverWithClickInstead = false;
			beanTwo.UnhoverWithClickInstead = false;
			beanOne.UnhoverWithJavascriptClickInstead = true;
			beanTwo.UnhoverWithJavascriptClickInstead = true;

			Assert.AreEqual(beanOne, beanTwo, "Two beans should be equal if they have the same container and unhover " + 
				"element references and same non-default JavaScript click action unhover workaround value");
			
			beanOne = new HoverControlBean();
			beanTwo = new HoverControlBean();

			beanOne.HoverWithJavascript = true;
			beanTwo.HoverWithJavascript = true;
			beanOne.UnhoverWithJavascript = true;
			beanTwo.UnhoverWithJavascript = true;
			beanOne.ClickInsteadOfHover = true;
			beanTwo.ClickInsteadOfHover = true;
			beanOne.ClickWithJavascriptInsteadOfHover = true;
			beanTwo.ClickWithJavascriptInsteadOfHover = true;
			beanOne.UnhoverWithClickInstead = true;
			beanTwo.UnhoverWithClickInstead = true;
			beanOne.UnhoverWithJavascriptClickInstead = true;
			beanTwo.UnhoverWithJavascriptClickInstead = true;

			Assert.AreEqual(beanOne, beanTwo, "Two beans should be equal if they have null container and unhover " + 
				"element references and same non-default JavaScript hover workaround values");
		}

		[Test]
		[Category("ToStringCallsBaseTestGroup")]
		[Category("BeanToStringCallsBaseTestGroup")]
		[Category("HoverControlBeanToStringCallsBaseTestGroup")]
		public void ShouldCallBaseForToString() {
			Mock<IWebElement> mockElement = new Mock<IWebElement>();

			IHoverControlBean bean = new HoverControlBean();
			bean.ContentContainer = mockElement.Object;

			Assert.AreEqual("HoverControlBean(ContentContainerBean(LoadableBean(Driver: null, LoadTimeout: 30), " + 
				"ContentContainer: " + mockElement.Object.ToString() + "), UnhoverElement: null, " + 
				"HoverWithJavascript: False, UnhoverWithJavascript: False, ClickInsteadOfHover: False, " + 
				"ClickWithJavascriptInsteadOfHover: False, UnhoverWithClickInstead: False, " + 
				"UnhoverWithJavascriptClickInstead: False)", bean.ToString());
		}

		[Test]
		[Category("HashCodeCallsSuperTestGroup")]
		[Category("BeanHashCodeCallsSuperTestGroup")]
		[Category("HoverControlBeanHashCodeCallsSuperTestGroup")]
		public void ShouldCallBaseForHashCode() {
			IHoverControlBean bean = new HoverControlBean();
			IHoverControlBean beanToCompare = new HoverControlBean();

			Assert.AreEqual(bean.GetHashCode(), beanToCompare.GetHashCode(), "Hash codes for bean which have not had " + 
				"setters called should be equal, but are not: " + bean.ToString() + ", " + beanToCompare.ToString());

			Mock<IWebElement> mockElement = new Mock<IWebElement>();

			bean.ContentContainer = mockElement.Object;
			Assert.AreNotEqual(bean.GetHashCode(), beanToCompare.GetHashCode(), "Hash codes for bean which have " + 
				"different values for their container element fields should not be equal, but are: " + 
				bean.ToString() + ", " + beanToCompare.ToString());

			beanToCompare.ContentContainer = mockElement.Object;
			Assert.AreEqual(bean.GetHashCode(), beanToCompare.GetHashCode(), "Hash codes for bean which have the " + 
				"same container element should be equal, but are not: " + bean.ToString() + ", " + 
				beanToCompare.ToString());
		}

		[Test]
		[Category("ControlBeanExtensionMethodsTestGroup")]
		public void ShouldReturnTrueForIsHoverControl() {
			IHoverControlBean bean = new HoverControlBean();
			Assert.IsTrue(bean.IsHoverControl());
		}	

		[Test]
		[Category("ControlBeanExtensionMethodsTestGroup")]
		public void ShouldReturnFalseForIsClickControl() {
			IHoverControlBean bean = new HoverControlBean();
			Assert.IsFalse(bean.IsClickControl());
		}

		[Test]
		[Category("ControlBeanExtensionMethodsTestGroup")]
		public void ShouldReturnFalseForIsHoverAndClickControl() {
			IHoverControlBean bean = new HoverControlBean();
			Assert.IsFalse(bean.IsHoverAndClickControl());
		}
	}
}

